File: backend/pkg/utils/validation_test.go
package utils

import (
	"strings"
	"testing"
)

func TestValidateEmail(t *testing.T) {
	tests := []struct {
		email string
		want  bool
	}{
		{"user@example.com", true},
		{"user.name@example.com", true},
		{"user+tag@example.com", true},
		{"user@sub.example.com", true},
		{"", false},
		{"no-at-sign", false},
		{"@no-local.com", false},
		{"user@", false},
		{"user@.com", false},
		{"user@com", false},
		{"user@example", false},
		{string(make([]byte, 300)), false}, // too long
	}

	for _, tt := range tests {
		err := ValidateEmail(tt.email)
		got := err == nil
		if got != tt.want {
			t.Errorf("ValidateEmail(%q) = %v, want %v (err: %v)", tt.email, got, tt.want, err)
		}
	}
}

func TestValidateUsername(t *testing.T) {
	tests := []struct {
		username string
		want     bool
	}{
		{"john", true},
		{"john_doe", true},
		{"john-doe", true},
		{"user123", true},
		{"ab", false},   // too short
		{"admin", false}, // reserved
		{"system", false}, // reserved
		{"_john", false}, // starts with special
		{"-john", false}, // starts with special
		{"john--doe", false}, // consecutive special chars
		{"john__doe", false}, // consecutive special chars
		{"john doe", false},  // contains space
		{"", false},          // empty
	}

	// Add a long username test
	longUsername := ""
	for i := 0; i < 33; i++ {
		longUsername += "a"
	}
	tests = append(tests, struct {
		username string
		want     bool
	}{longUsername, false})

	for _, tt := range tests {
		err := ValidateUsername(tt.username)
		got := err == nil
		if got != tt.want {
			t.Errorf("ValidateUsername(%q) = %v, want %v (err: %v)", tt.username, got, tt.want, err)
		}
	}
}

func TestValidatePassword(t *testing.T) {
	tests := []struct {
		name     string
		password string
		username string
		wantErr  bool
	}{
		{"valid strong", "MyPass123!", "user", false},
		{"valid without special", "MyPass123", "user", false},
		{"too short", "Ab1", "user", true},
		{"too long", "Aa1" + strings.Repeat("x", 62), "user", true},
		{"no uppercase - allowed (NIST)", "mypassword1", "user", false},
		{"no lowercase - allowed (NIST)", "MYPASSWORD1", "user", false},
		{"no number - allowed (NIST)", "MyPassword", "user", false},
		{"long passphrase (NIST)", "correcthorsebatterystaple", "user", false},
		{"common password", "password", "user", true},
		{"all same chars", "AAAAAAAAAAA", "user", true},
		{"contains username", "MyUser123", "myuser", true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := ValidatePassword(tt.password, tt.username)
			gotErr := err != nil
			if gotErr != tt.wantErr {
				t.Errorf("ValidatePassword(%q, %q) error = %v, wantErr %v", tt.password, tt.username, gotErr, tt.wantErr)
			}
		})
	}
}

func TestIsDisposableEmail(t *testing.T) {
	tests := []struct {
		email string
		want  bool
	}{
		{"user@mailinator.com", true},
		{"user@guerrillamail.com", true},
		{"user@10minutemail.com", true},
		{"user@gmail.com", false},
		{"user@company.com", false},
		{"user@example.com", false},
	}

	for _, tt := range tests {
		got := IsDisposableEmail(tt.email)
		if got != tt.want {
			t.Errorf("IsDisposableEmail(%q) = %v, want %v", tt.email, got, tt.want)
		}
	}
}

func TestGenerateUsernameFromEmail(t *testing.T) {
	tests := []struct {
		email string
		want  string
	}{
		{"john@gmail.com", "john"},
		{"john.doe@gmail.com", "johndoe"},
		{"john+tag@gmail.com", "john"},
		{"ab@x.com", "user"},           // too short (< 3)
		{"123user@gmail.com", "user_123user"}, // starts with non-alpha
	}

	for _, tt := range tests {
		got := GenerateUsernameFromEmail(tt.email)
		if got != tt.want {
			t.Errorf("GenerateUsernameFromEmail(%q) = %q, want %q", tt.email, got, tt.want)
		}
	}
}

func TestSanitizeHTML(t *testing.T) {
	tests := []struct {
		input string
		want  string
	}{
		{"<script>alert(1)</script>", "&lt;script&gt;alert(1)&lt;/script&gt;"},
		{`"onclick"='xss'`, "&quot;onclick&quot;=&#x27;xss&#x27;"},
		{"normal text", "normal text"},
		{"a & b", "a &amp; b"},
	}

	for _, tt := range tests {
		got := SanitizeHTML(tt.input)
		if got != tt.want {
			t.Errorf("SanitizeHTML(%q) = %q, want %q", tt.input, got, tt.want)
		}
	}
}

File: frontend/src/lib/__tests__/validation.test.ts
import { describe, it, expect } from 'vitest'
import {
  validateEmail,
  validateUsername,
  validatePassword,
  PasswordStrength,
} from '../validation'

describe('validateEmail', () => {
  it('accepts valid emails', () => {
    expect(validateEmail('user@example.com').valid).toBe(true)
    expect(validateEmail('user.name@example.com').valid).toBe(true)
    expect(validateEmail('user+tag@example.com').valid).toBe(true)
  })

  it('rejects empty email', () => {
    expect(validateEmail('').valid).toBe(false)
  })

  it('rejects email without @', () => {
    expect(validateEmail('no-at-sign').valid).toBe(false)
  })

  it('rejects email with short domain', () => {
    expect(validateEmail('user@a').valid).toBe(false)
  })

  it('rejects overly long email', () => {
    const long = 'a'.repeat(250) + '@x.com'
    expect(validateEmail(long).valid).toBe(false)
  })
})

describe('validateUsername', () => {
  it('accepts valid usernames', () => {
    expect(validateUsername('john').valid).toBe(true)
    expect(validateUsername('john_doe').valid).toBe(true)
    expect(validateUsername('user123').valid).toBe(true)
  })

  it('rejects too short username', () => {
    expect(validateUsername('ab').valid).toBe(false)
  })

  it('rejects reserved usernames', () => {
    expect(validateUsername('admin').valid).toBe(false)
    expect(validateUsername('system').valid).toBe(false)
  })

  it('rejects consecutive special chars', () => {
    expect(validateUsername('john--doe').valid).toBe(false)
    expect(validateUsername('john__doe').valid).toBe(false)
  })

  it('rejects username starting with special char', () => {
    expect(validateUsername('_john').valid).toBe(false)
    expect(validateUsername('-john').valid).toBe(false)
  })
})

describe('validatePassword', () => {
  it('rejects empty password', () => {
    expect(validatePassword('').error).toBeDefined()
  })

  it('rejects password shorter than 8 chars', () => {
    const result = validatePassword('Ab1')
    expect(result.error).toBeDefined()
    expect(result.strength).toBe(PasswordStrength.Weak)
  })

  it('rejects password longer than 64 chars', () => {
    const result = validatePassword('A'.repeat(65) + '1a')
    expect(result.error).toBeDefined()
  })

  it('rejects common weak passwords', () => {
    expect(validatePassword('password').error).toBeDefined()
    expect(validatePassword('12345678').error).toBeDefined()
    expect(validatePassword('qwerty').error).toBeDefined()
  })

  it('rejects all same characters', () => {
    expect(validatePassword('AAAAAAAAAAA').error).toBeDefined()
  })

  it('accepts long passphrases without mixed case (NIST)', () => {
    const result = validatePassword('correcthorsebatterystaple')
    expect(result.error).toBeUndefined()
  })

  it('scores strong passwords highly', () => {
    const result = validatePassword('MyStr0ng!Passw0rd#2024')
    expect(result.error).toBeUndefined()
    expect(result.strength).toBeGreaterThanOrEqual(PasswordStrength.Good)
  })
})
